File: src/main/java/com/rimlang/rim/syntax/ArgsExpression.java
package com.rimlang.rim.syntax;

import com.rimlang.rim.lexer.StringToken;
import com.rimlang.rim.lexer.Token;
import com.rimlang.rim.lexer.TokenType;
import com.rimlang.rim.translation.Translator;
import com.rimlang.rim.util.TokenUtil;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ArgsExpression extends Expression {

    private final List<Argument> args = new ArrayList<>();

    public ArgsExpression(List<Token> tokens) throws RimSyntaxException {
        for (List<Token> arg : TokenUtil.split(tokens, TokenType.COMMA)) {
            if (arg.get(0).getType() != TokenType.ID) {
                throw new RimSyntaxException("Argument must have identifier", arg.get(0).getLine());
            }
            StringToken identifier = (StringToken) arg.get(0);
            if (arg.get(1).getType() != TokenType.COLON) {
                throw new RimSyntaxException("Argument must have a separating colon between the identifier and the argument type", arg.get(1).getLine());
            }
            if (arg.get(2).getType() != TokenType.TYPE) {
                System.out.println(arg);
                throw new RimSyntaxException("Argument must have type", arg.get(2).getLine());
            }
            StringToken type = (StringToken) arg.get(2);
            if (arg.size() > 4 && arg.get(3).getType() == TokenType.ASSIGNMENT_OPERATOR) {
                args.add(new Argument(type, identifier, TokenUtil.sub(arg, 4)));
            } else {
                args.add(new Argument(type, identifier, null));
            }
        }
    }

    public String translate() {
        return args.stream().map(Argument::translate).collect(Collectors.joining(", "));
    }

    public List<Argument> getArgs() {
        return args;
    }

    public static class Argument {
        private final StringToken identifier;
        private final StringToken type;
        private final List<Token> defaultValue;

        public Argument(StringToken type, StringToken identifier, @Nullable List<Token> defaultValue) {
            this.type = type;
            this.identifier = identifier;
            this.defaultValue = defaultValue;
        }

        public String translate() {
            return Translator.toJavaType(type) + " " + Translator.getCamelCase(identifier.getValue());
        }
    }

}

File: src/main/java/com/rimlang/rim/syntax/ArgsExpression.kt
package com.rimlang.rim.syntax

import com.rimlang.rim.lexer.StringToken
import com.rimlang.rim.lexer.Token
import com.rimlang.rim.lexer.TokenType
import com.rimlang.rim.translation.Context
import com.rimlang.rim.translation.Translator
import com.rimlang.rim.util.camelCase
import com.rimlang.rim.util.split
import com.rimlang.rim.util.sub

class ArgsExpression(tokens: List<Token>) : Expression() {
    val args = mutableListOf<Argument>()

    init {
        for (arg in tokens.split(TokenType.COMMA)) {
            if (arg[0].type !== TokenType.ID) {
                throw RimSyntaxException("Argument must have identifier", arg[0].line)
            }
            val identifier = arg[0] as StringToken
            if (arg[1].type !== TokenType.COLON) {
                throw RimSyntaxException(
                    "Argument must have a separating colon between the identifier and the argument type",
                    arg[1].line
                )
            }
            if (arg[2].type !== TokenType.TYPE) {
                throw RimSyntaxException("Argument must have type", arg[2].line)
            }
            val type = arg[2] as? StringToken ?: continue
            args += if (arg.size > 4 && arg[3].type === TokenType.ASSIGNMENT_OPERATOR) {
                Argument(type, identifier, sub(arg, 4))
            } else {
                Argument(type, identifier)
            }
        }
    }

    data class Argument(
        val type: StringToken,
        val identifier: StringToken,
        val defaultValue: List<Token> = listOf()
    ) {
        fun translate(): String {
            return Translator.toJavaType(type) + " " + identifier.value.camelCase()
        }
    }

    override fun translate(context: Context): String {
        return args.joinToString { it.translate() }
    }

}